AOC [23.09](https://adventofcode.com/2023/day/9) Mirage Maintenance

Input is data from an environmental sensor. Each line contains the history of a single variable.  Figure out an equation to predict next value for each line.

Procedure: 
1. for each sequence of numbers, create a new sequence that is the difference between the numbers
* 1, 3, 6, 10, 15, 21 --> 2, 3, 4, 5, 6
1. Keep doing this until all the differences are zeros:
* 1, 3, 6, 10, 15, 21 --> 2, 3, 4, 5, 6 --> 1, 1, 1 --> 0, 0, 0
1. start with zero and add the last digit in each line:
* 0 + 1 = 1
* 1 + 6 = 7
* 7 + 21 = 28

Find the sum of all the lists in the input.

Solution:
1. write a function that takes a list of numbers and returns the next value in the list
1. the function should do the process described above (create a series of lists)


In [1]:
test_data = '''0 3 6 9 12 15
1 3 6 10 15 21
10 13 16 21 30 45'''

data = [line for line in test_data.split('\n')]
data = [list(map(int, x.split())) for x in data]

data

[[0, 3, 6, 9, 12, 15], [1, 3, 6, 10, 15, 21], [10, 13, 16, 21, 30, 45]]

In [18]:
[list(map(int, x.split())) for x in line for line in test_data.split('\n')]

NameError: name 'line' is not defined

In [2]:
def find_diffs(sequence): # input list of values; returns list of differences in values
    diffs = []
    for i in range(len(sequence)-1):
        diffs.append(sequence[i+1] - sequence[i])
    return diffs


Thinking about the problem a bit more, I don't need to save all the different lists that are produced (as the assignment suggests). Rather, I just need to add the final element of each list to the final element of the current list. This is much easier!

In [3]:
def find_diffs(sequence): # input list of values; returns list of differences in values
    diffs = []
    for i in range(len(sequence)-1):
        diffs.append(sequence[i+1] - sequence[i])
    return diffs


last_val = 10**6
curr_seq = [10 , 13  ,16 , 21 , 30,  45 ]
next_val = curr_seq[-1] # get the last value in the curr sequence
print(curr_seq,last_val, next_val)

while last_val > 0:
    diffs = find_diffs(curr_seq) # find the last val in the list of differences
    last_val = diffs[-1]
    next_val += last_val
    curr_seq = diffs
    print(curr_seq,last_val, next_val)



[10, 13, 16, 21, 30, 45] 1000000 45
[3, 3, 5, 9, 15] 15 60
[0, 2, 4, 6] 6 66
[2, 2, 2] 2 68
[0, 0] 0 68


In [4]:

data = [line for line in test_data.split('\n')]
data = [list(map(int, x.split())) for x in data]

data

def find_diffs(sequence): # input list of values; returns list of differences in values
    diffs = []
    for i in range(len(sequence)-1):
        diffs.append(sequence[i+1] - sequence[i])
    return diffs

tally = 0
for curr_seq in data:
    last_val = 10**6
    next_val = curr_seq[-1]
    print(curr_seq)
    while last_val > 0:
        diffs = find_diffs(curr_seq) # find the last val in the list of differences
        last_val = diffs[-1]
        next_val += last_val
        curr_seq = diffs
    print(f"next element = {next_val}")
    tally += next_val
print(tally)
        


[0, 3, 6, 9, 12, 15]
next element = 18
[1, 3, 6, 10, 15, 21]
next element = 28
[10, 13, 16, 21, 30, 45]
next element = 68
114


Test data works fine. Let's try production data...

In [27]:
with open ('2309input.txt') as f_in:
    data = [line for line in f_in.read().split('\n')]

#data = [line for line in test_data.split('\n')]
data = [list(map(int, x.split())) for x in data]

data

def find_diffs(sequence): # input list of values; returns list of differences in values
    diffs = []
    for i in range(len(sequence)-1):
        diffs.append(sequence[i+1] - sequence[i])
    return diffs

tally = 0
for curr_seq in data:
    #last_val = 10**10     
    next_val = curr_seq[-1] # last value in the current list of numbers
    #print(curr_seq, next_val)
    while set(curr_seq) != {0}:
        diffs = find_diffs(curr_seq) # find the last val in the list of differences
        #last_val = diffs[-1]
        next_val += diffs[-1]
        curr_seq = diffs
     #   print(curr_seq, next_val)
    tally += next_val

print(f"{tally:,},  {tally}")


1,877,825,184,  1877825184


Run #1:
* $1,879,448,560$ is too high a value.

Run #2 - change the test from
`last_val > 0` to `last_val != 0`
* $1,877,825,200$ is still too high

Run #3 - since the criteria is `all numbers in list = 0`, change the test to: `while sum(curr_seq) != 0`
* $1,877,825,182$ is too *low*

Run #4 - `sum(curr_seq)` isn't the right test: `sum([-1,1,0,0]) = 1`. So change test to `set(curr_seq) != {0}`. This generated teh correct result: $1,877,825,184$.



Part 2: Same process but with the first value in the sequence. Given the way I've set this up, this should be quick and easy. . . 

In [39]:
with open ('2309input.txt') as f_in:
    data = [line for line in f_in.read().split('\n')]

#data = [line for line in test_data.split('\n')]
data = [list(map(int, x.split())) for x in data]

data

def find_diffs(sequence): # input list of values; returns list of differences in values
    diffs = []
    for i in range(len(sequence)-1):
        diffs.append(sequence[i] - sequence[i+1])
    return diffs

tally = 0
for curr_seq in data:   
    next_val = curr_seq[0] # first value in the current list of numbers
    #print(next_val, curr_seq)
    while set(curr_seq) != {0}:
        diffs = find_diffs(curr_seq) # find the last val in the list of differences
        curr_seq = diffs
        next_val += diffs[0]   
    #    print(next_val, curr_seq)
    tally += next_val 

print(f"{tally:,},  {tally}")

1,108,  1108


Got it correct on the first production run. Boom!
